src/cargo/sources/config.rs

```diff
@@ -104,19 +104,25 @@ impl<'cfg> SourceConfigMap<'cfg> {
         }
         let new_src = new_id.load(self.config);
         let old_src = id.load(self.config);
-        if new_src.supports_checksums() != old_src.supports_checksums() {
-            let (supports, no_support) = if new_src.supports_checksums() {
-                (name, orig_name)
-            } else {
-                (orig_name, name)
-            };
+        if !new_src.supports_checksums() && old_src.supports_checksums() {
             bail!("\
 cannot replace `{orig}` with `{name}`, the source `{supports}` supports \
 checksums, but `{no_support}` does not
 
 a lock file compatible with `{orig}` cannot be generated in this situation
-", orig = orig_name, name = name, supports = supports, no_support = no_support);
+", orig = orig_name, name = name, supports = orig_name, no_support = name);
```

Comment:
We don't need `supports` and `no_support` keys any more, may just use `orig` and `name` (or `replaced`)